### ISSUE TITLE
Gradient of tf.py_func and how to define gradients for a custom op only with python

### ISSUE THREAD

**Issue by carpedm20** [tensorflow/tensorflow#1095]
I know about `tf.py_func` which I can use a python code as an op but how gradient is calculated with it?

Also, is there anyway I can set the gradient only with python? I think one way might be adding a custom op that calculate gradients into an array and use the array as an argument of `apply_gradients`.

For example:

    x = tf.Variable(3.) # this is an custom op
    y = tf.Variable(4.)
    loss = 13 - x * y

    opt = tf.train.GradientDescentOptimizer(0.1)
    [(y_grad, y_val)] = opt.compute_gradients(loss, [y])
    clipped_grads_and_vars = [(x * y, x), (y_grad, y_val)] # custom gradient is x * y

    optim = opt.apply_gradients(clipped_grads_and_vars)

Is this safe way to apply custom gradients only with python?

**Comment by fchollet** [comment]
You can use the Python decorator `@tf.RegisterGradient` to add a gradient to a Python op while only touching Python code. https://www.tensorflow.org/versions/v0.6.0/api_docs/python/framework.html#RegisterGradient

**Comment by ibab** [comment]
It seems that using `@tf.RegisterGradient` doesn't work, as it is specific to a certain `op.type` (which is always `PyFunc` here), and not to `op.name`.
The only way to set `op.type` seems to be to define a custom C++ op.
Did I miss something, or can someone confirm this?
Would be nice if I could define custom gradients for my Python ops.

**Comment by zffchen78** [comment]
Here is a sketch you can you extend the current code to implement the gradient for py_func and all changes will be restricted to script_ops.py:

1.

**Comment by ibab** [comment]
@zffchen78: It seems that your sketch didn't make it into the post :confused:

Maybe you were suggesting to define a `tf.RegisterGradient('PyFunc')` that dispatches to a different  registry of gradients that is based on something other than `op.type` (maybe `op.name`, or maybe the underlying Python function?).

Entries in that registry could be defined with something like `tf.RegisterPyFuncGradient('MyCustomOp')`

That would indeed be straightforward to implement, but maybe you had something else in mind?

**Comment by zffchen78** [comment]
oops. sorry that I were distracted. But yes, you are welcome to extend the code to support gradient of a pyfunc. What I have in mind may be slightly different from your sketch. I'm thinking you can change
tf.py_func(lambda, ...) to take an extra argument grad_func=lambda. In the py_func.py's registry, keep lambda and grad_lambda keyed by the unique token. In the PyFuncGrad(op), dispatch to grad_lambda based on op.attr['token'].

**Comment by girving** [comment]
I don't think `py_func` should have any special functionality here.  You can already do this with a combination of `RegisterGradient` and `Graph.gradient_override_map`.  It's ugly, but if we want to make it less ugly we should clean up gradient overriding for all ops, not just `py_func`.

**Comment by ibab** [comment]
Ah, I didn't know about `Graph.gradient_override_map`.
It could be used to implement the API @zffchen78 suggested like this:
```python
def py_func(func, grad=None):
  tf.RegisterGradient("CustomPyFunc123")(grad) # Generate a unique name here
  g = tf.get_default_graph()
  with g.gradient_override_map({"PyFunc": "CustomPyFunc123"}):
    return old_py_func(func)
```
But having to generate a unique name is quite ugly.
Maybe a possible future gradient overriding method won't have that requirement.
It would be simple if `gradient_override_map` used a function directly like this:
```
def grad_func(op, grad):
  # ...
with g.gradient_override_map({"PyFunc": grad_func}):
  # ...
```

**Comment by girving** [comment]
Agreed, `gradient_override_map` is pretty ugly.  Separate issue from py_func, but so far I don't think we have either an internal or external bug tracking that ugliness. :)

**Comment by harpone** [comment]
In case some other n00b like me stumbles here wondering how to practically implement the py_func grad, I came up with an explicit example (after some struggling) which I posted [here](https://gist.github.com/harpone/3453185b41d8d985356cbe5e57d67342).

The "unique" name generation is a bit hacky...

**Comment by chenqifeng22** [comment]
@harpone Thanks for sharing the code! It really helps me. I was searching for such a solution for a long time.

**Comment by ruppeshnalwaya1993** [comment]
Hi! I wrote a custom binarizer inspired from @harpone 's code. But I am facing problem if nan output cost when using it in two copies of same network simultaneously. Detailed question is here on stackoverflow : http://stackoverflow.com/questions/40735766/tensorflow-custom-operation-used-in-two-networks-simultaneously-produces-nan

Could anyone please answer my query?

**Comment by adler-j** [comment]
Regarding the method used by @harpone, the example given only shows how to do this if the gradient can be defined in pure TensorFlow. How would you go ahead if both the function and its gradient should be computed with e.g. numpy?

**Comment by IFLED** [comment]
@adler-j you can define op using `tf.py_func` that have additional outputs: derivatives of output with respect to inputs. Next, when you will define gradient of the new op, you can define gradient with respect to inputs it pure TensorFlow (using additional outputs of the new op).

I posted example [here](https://gist.github.com/harpone/3453185b41d8d985356cbe5e57d67342#gistcomment-2070556).

**Comment by CassieMai** [comment]
Hello, everybody. I am trying to embed a python function into my tensorflow network code using tf.py_func. But I find that my code doesn't go into my python function. That's strange. Do anyone have suggestion?

**Comment by IFLED** [comment]
Hello, @CassieMai. Could you provide some code? Or maybe describe in more details?
How did you find out that code doesn't go into python function?

**Comment by CassieMai** [comment]
Hello, @IFLED, thanks.  The layer definition and python function definition are as following. When my code was executed, it didn't go into myPyFunc at all. Did I miss anything?

def myLayer(self, input, name):
        with tf.variable_scope(name) as scope:
             d_num = input.get_shape()[-1]
             d = tf.reshape(input[0][:,:,1], [1,-1])
             e = tf.py_func(myPyFunc, [d], [tf.float32])
             e = tf.convert_to_tensor(e)
             return e

def myPyFunc(input):
        print "===Hahaha==="
        return 0

**Comment by adler-j** [comment]
@CassieMai that is not a [minimal working example](https://stackoverflow.com/help/mcve). Can you please provide the full code?

**Comment by IFLED** [comment]
```
import tensorflow as tf
import numpy as np

def myLayer(self, input, name):
        with tf.variable_scope(name) as scope:
             d_num = input.get_shape()[-1]
             d = tf.reshape(input[0][:,:,1], [1,-1])
             e = tf.py_func(myPyFunc, [d], [tf.float32])
             e = tf.convert_to_tensor(e)
             return e

def myPyFunc(input):
        print "===Hahaha==="
        return np.float32(0)

input_tensor = tf.placeholder(tf.float32, (2,2,2,2))
input_value = np.ones((2,2,2,2))

output_tensor = myLayer("asdf", input_tensor, "asdf")

with tf.Session() as sess:
    print sess.run(output_tensor, { input_tensor : input_value })
```

@CassieMai, this code produces following output:
```
===Hahaha===
[ 0.]
```

**Comment by CassieMai** [comment]
Thank you @IFLED @adler-j . It's ok to run this small code by inputting a custom input_value. Maybe there is something wrong for my input in my code. I am not familiar with tensor operations on tensor format.

**Comment by thanard** [comment]
How can I output a full Jacobean matrix to `grad_func`, which has size  |output| x |input|?

**Comment by adler-j** [comment]
Sure can, in my example above you replace

```
def _MySquareGrad(op, grad):
    x = op.inputs[0]
    return grad * 20 * x
```

with

```
def _MySquareGrad(op, grad):
    x = op.inputs[0]
    jacobian = ComputeJacobian(x)  # compute jacobian in x
    return np.dot(jacobian.T, grad)
```

with that said, this is in general inferior to an explicit computation.

**Comment by xisnu** [comment]
> def _MySquareGrad(op, grad):

I was wondering what exactly this **grad** is doing. I was trying to design a custom mean squared error. My loss computation is
```
def mse_loss(logits,labels):
    total=len(logits)
    loss=0.0
    #print(logits)
    #print(labels)
    for t in range(total):
        s=0.0
        for j in range(2):
           s=s+(logits[t][j]-labels[t][j])**2
        loss=loss+s
    #loss=loss/float(total)
    return loss
```
for this function I have defined a gradient function as
```
def _MeanSquareGrad(op, grad):
    x = op.inputs[0]
    y = op.inputs[1]
    return grad * (x-y),grad*0.01
```
I called it in main tensorflow body as
`loss=py_func(mse_loss,[logits,y],tf.float64,grad=_MeanSquareGrad)`
Now, the code is correctly finding the Loss and optimizer is also minimizing it. But I am not sure about

1. What is the purpose of using ` grad * (x-y)`. What is the value of grad?
2. I have send a dummy `grad*0.01` value back as I am sending two inputs as in `py_func(mse_loss,[logits,y],tf.float64,grad=_MeanSquareGrad)`. Why it is asking for two returns from grad procedure

**Comment by IFLED** [comment]
1. if you call `tf.gradients(F, x)` (and F depends on `loss`) then `grad` will be derivative of `F` with respect to `loss`. For example you can take `F = tf.sqrt(loss)`, print `grad` with different values of labels or logits.
2. `mse_loss` is a function that depends on two inputs (`logits`, `labels`), so `grad` procedure should return derivatives with respect to both inputs.

**Comment by raminhalavati** [comment]
Is there a document on what the 'grad' argument in the gradient operator do? (the |grad| in |_MeanSquareGrad(op, grad)|)

**Comment by IFLED** [comment]
What do you mean by document?

As far as I remember, [here](https://github.com/tensorflow/tensorflow/issues/1095#issuecomment-378592178) function `_MeanSquareGrad` is used to define gradent of `loss` op and param `grad` is a tensor that golds gradients of "some global function" with respect to output of `loss` op. See [my previous comment](https://github.com/tensorflow/tensorflow/issues/1095#issuecomment-384059102).

**Comment by raminhalavati** [comment]
Thank you, I was just asking if there is a detailed description about it somewhere, as I don't clearly get it from previous comments.
I have defined loss value by a py_func, and that's why I am overriding the gradient to define the actual gradient of it. Now I don't get what exactly |grad| means and how should I use it. I don't clearly get what you mean by "gradient of some global function with respect to outputs of loss op".
